tests/std/tests/P2387R3_pipe_support_for_user_defined_range_adaptors/test.cpp

```diff
@@ -0,0 +1,206 @@
+        return {};
+    }
+};
+static_assert(CanPipe_R<TestRange&, RangeAdaptorClosureParameterRefQualTest, char>);
+static_assert(CanPipe_R<TestRange&&, RangeAdaptorClosureParameterRefQualTest, short>);
+static_assert(CanPipe_R<const TestRange&, RangeAdaptorClosureParameterRefQualTest, unsigned char>);
+static_assert(CanPipe_R<const TestRange&&, RangeAdaptorClosureParameterRefQualTest, unsigned short>);
+
+using FirstIdentityThenMemberRefQualTest = decltype(
+    IdentityRangeAdaptorClosure{} | RangeAdaptorClosureMemberRefQualTest{});
+static_assert(CanPipe_R<TestRange, FirstIdentityThenMemberRefQualTest&, char>);
+static_assert(CanPipe_R<TestRange, FirstIdentityThenMemberRefQualTest&&, short>);
+static_assert(CanPipe_R<TestRange, FirstIdentityThenMemberRefQualTest const&, unsigned char>);
+static_assert(CanPipe_R<TestRange, FirstIdentityThenMemberRefQualTest const&&, unsigned short>);
+
+struct MoveOnlyRangeAdaptorClosure : std::ranges::range_adaptor_closure<MoveOnlyRangeAdaptorClosure> {
+    MoveOnlyRangeAdaptorClosure()                                   = default;
+    MoveOnlyRangeAdaptorClosure(const MoveOnlyRangeAdaptorClosure&) = delete;
+    MoveOnlyRangeAdaptorClosure(MoveOnlyRangeAdaptorClosure&&)      = default;
+
+    void operator()(const TestRange&) {}
+};
+static_assert(CanPipe<TestRange, MoveOnlyRangeAdaptorClosure>);
+static_assert(CanPipe<TestRange, MoveOnlyRangeAdaptorClosure&>);
+static_assert(CanPipe<MoveOnlyRangeAdaptorClosure, IdentityRangeAdaptorClosure>);
+static_assert(CanPipe<IdentityRangeAdaptorClosure, MoveOnlyRangeAdaptorClosure>);
+static_assert(!CanPipe<MoveOnlyRangeAdaptorClosure&, IdentityRangeAdaptorClosure>);
+static_assert(!CanPipe<IdentityRangeAdaptorClosure, MoveOnlyRangeAdaptorClosure&>);
+
+class TimesTwoAdaptor : public std::ranges::range_adaptor_closure<TimesTwoAdaptor> {
+public:
+    template <std::ranges::range R>
+    constexpr auto operator()(R&& range) const {
+        return std::forward<R>(range) | std::views::transform([](auto x) { return x * 2; });
+    }
+};
+
+class DividedByTwoAdaptor : public std::ranges::range_adaptor_closure<DividedByTwoAdaptor> {
+public:
+    template <std::ranges::range R>
+    constexpr auto operator()(R&& range) const {
+        return std::forward<R>(range) | std::views::transform([](auto x) { return x / 2; });
+    }
+};
+
+constexpr bool test_constexpr() {
+    const std::array<int, 3> numbers{1, 2, 3};
+    const std::array<int, 3> numbers_times_two{2, 4, 6};
+    assert(std::ranges::equal(numbers_times_two, numbers | TimesTwoAdaptor{}));
+    assert(std::ranges::equal(numbers, (numbers | TimesTwoAdaptor{}) | DividedByTwoAdaptor{}));
+    assert(std::ranges::equal(numbers, numbers | (TimesTwoAdaptor{} | DividedByTwoAdaptor{})));
+    assert(std::ranges::equal(numbers_times_two,
+        (numbers | TimesTwoAdaptor{})
+            | (TimesTwoAdaptor{} | TimesTwoAdaptor{} | DividedByTwoAdaptor{} | DividedByTwoAdaptor{})));

```

Comment:
The test should additionally verify that mixing standard range adaptor closure objects and user-defined range adaptor objects works correctly.

Something like

```c++
     auto pipeline = TimesTwoAdaptor{} | std::views::transform([](auto x) { return x / 2; });
     assert(std::ranges::equal(numbers, numbers | pipeline));
```
or simply
```c++
     assert(std::ranges::equal(numbers, numbers | (TimesTwoAdaptor{} | std::views::transform([](auto x) { return x / 2; }))));
```

Reply:
Damn, I totally missed that functionality. Indeed, with my implementation you can't mix standard range adaptor closures and user-defined ones. I will reiterate on this and find a solution.

Reply:
I fixed the issue. I _assume_ that my changes don't break ABI.